src/qps/interface/qps.c

```diff
@@ -683,14 +681,19 @@ PetscErrorCode QPSGetConvergenceContext(QPS qps,void **ctx)
 .seealso: QPSSetConvergenceTest(), QPSSetTolerances(), QPSConvergedSkip(), KSPConvergedReason, QPSGetConvergedReason(),
           QPSConvergedDefaultCreate(), QPSConvergedDefaultDestroy()
 @*/
-PetscErrorCode QPSConvergedDefault(QPS qps,QP qp,PetscInt i,PetscReal rnorm,KSPConvergedReason *reason,void *ctx)
+PetscErrorCode QPSConvergedDefault(QPS qps,KSPConvergedReason *reason,void *ctx)
 {
   QPSConvergedDefaultCtx *cctx = (QPSConvergedDefaultCtx*) ctx;
+  QP qp = qps->solQP;
+  PetscInt i = qps->iteration;
+  PetscReal rnorm = qps->rnorm;
 
   PetscFunctionBegin;
   PetscValidHeaderSpecific(qps,QPS_CLASSID,1);
+  PetscValidHeaderSpecific(qp,QP_CLASSID,2);
   *reason = KSP_CONVERGED_ITERATING;
 
+  if (!qps->setupcalled) FLLOP_SETERRQ(((PetscObject) qps)->comm, PETSC_ERR_ARG_WRONGSTATE, "QPSSetUp() not yet called");
   if (!cctx->setup_called) {
     TRY( QPSConvergedDefaultSetUp(ctx,qps) );
   }
```

Comment:
I can't comment on the line bellow, but this `if` should be swapped with the following one.

Reply:
yes